controller-server/src/main/java/com/yahoo/vespa/hosted/controller/versions/VersionStatus.java

```diff
@@ -152,29 +152,33 @@ public static VersionStatus compute(Controller controller) {
         return new VersionStatus(versions);
     }
 
-    private static ListMultimap<Version, HostName> findSystemApplicationVersions(Controller controller) {
-        ListMultimap<Version, HostName> versions = ArrayListMultimap.create();
-        for (ZoneApi zone : controller.zoneRegistry().zones().controllerUpgraded().zones()) {
-            for (SystemApplication application : SystemApplication.all()) {
-                List<Node> eligibleForUpgradeApplicationNodes = controller.serviceRegistry().configServer().nodeRepository()
-                        .list(zone.getId(), application.id()).stream()
-                        .filter(SystemUpgrader::eligibleForUpgrade)
-                        .collect(Collectors.toList());
-                if (eligibleForUpgradeApplicationNodes.isEmpty())
-                    continue;
-
-                boolean configConverged = application.configConvergedIn(zone.getId(), controller, Optional.empty());
+    private static NodeVersions findSystemApplicationVersions(Controller controller) {
+        var nodeVersions = controller.versionStatus().systemVersion()
+                                     .map(VespaVersion::nodeVersions)
+                                     .orElse(NodeVersions.EMPTY);
+        var hostnames = new HashSet<HostName>();
+        for (var zone : controller.zoneRegistry().zones().controllerUpgraded().zones()) {
+            for (var application : SystemApplication.all()) {
+                var nodes = controller.serviceRegistry().configServer().nodeRepository()
+                                      .list(zone.getId(), application.id()).stream()
+                                      .filter(SystemUpgrader::eligibleForUpgrade)
+                                      .collect(Collectors.toList());
+                if (nodes.isEmpty()) continue;
+                var configConverged = application.configConvergedIn(zone.getId(), controller, Optional.empty());
                 if (!configConverged) {
-                    log.log(LogLevel.WARNING, "Config for " + application.id() + " in " + zone.getId() + " has not converged");
+                    log.log(LogLevel.WARNING, "Config for " + application.id() + " in " + zone.getId() +
+                                              " has not converged");
                 }
-                for (Node node : eligibleForUpgradeApplicationNodes) {
+                var now = controller.clock().instant();
+                for (var node : nodes) {
                     // Only use current node version if config has converged
-                    Version nodeVersion = configConverged ? node.currentVersion() : controller.systemVersion();
-                    versions.put(nodeVersion, node.hostname());
+                    Version version = configConverged ? node.currentVersion() : controller.systemVersion();
+                    nodeVersions = nodeVersions.with(new NodeVersion(node.hostname(), version, now));
```

Comment:
Probably not an issue, but this makes runtime scale with the square of the number of nodes which have upgraded. Just storing the full `NodeVersion` object instead of the `hostnames`, and then creating the `NodeVersion` at the end, is a cheaper alternative.

Reply:
Yes, I considered this as well. Will find a better approach just in case.